### DISCUSSION TITLE
Piping output of executable with UTF-16 output

### DISCUSSION THREAD

**Discussion by Shayan-To** [PowerShell/PowerShell#17163] [Q&A] [answered]
I have problems piping the output of `wsl --list --quiet --running` to other commands. Here's an example:

```
PS> wsl --list --quiet --running | % { "'$_'" }
'Dist1'
''
'Dist2'
''
''
```

I inspected the output even more, and I found out that every other character in the strings is a null character. It seems like the output is encoded in UTF-16, and I could fix the problem by doing this:

```
PS> wsl --list --quiet --running | % {$_ -Replace '\0', ''} | ? Length -NE 0 | % { "'$_'" }
'Dist1'
'Dist2'
```

Is there a way to tell PowerShell to actually treat the output as UTF-16?

(Here's the proof that every other character is a null character:)

```
PS> wsl --list --quiet --running | % {$_.ToCharArray() -Join '|'}
D||i||s||t||1|

|D||i||s||t||2|


```

**Comment by jborean93** [comment, marked as answer]
To control what encoding PowerShell reads the raw stdout with you can set `[Console]::OutputEncoding`.

```powershell
$origEncoding = [Console]::OutputEncoding
try {
    [Console]::OutputEncoding = [System.Text.Encoding]::Unicode
    $output = wsl ...
}
finally {
    [Console]::OutputEncoding = $origEncoding
}
```

The downside of this approach is that it's a global setting for the process. There's no PowerShell specific variable that can be set at the Runspace/scope level like `$ErrorActionPreference` so you need to be careful if doing things in parallel.
